dashboard/lib/services/curriculum_pdfs/lesson_plans.rb

```diff
@@ -39,8 +39,13 @@ def get_lesson_plan_url(lesson, student_facing=false)
 
         # Check S3 to see if we've already generated a PDF for the given lesson.
         def lesson_plan_pdf_exists_for?(lesson, student_facing=false)
-          pathname = get_lesson_plan_pathname(lesson, student_facing)
-          AWS::S3.cached_exists_in_bucket?(S3_BUCKET, pathname.to_s)
+          pathname = get_lesson_plan_pathname(lesson, student_facing).to_s
+          cache_key = "CurriculumPdfs/LessonPlans/pdf_exists/#{pathname}"
+          return CDO.shared_cache.read(cache_key) if CDO.shared_cache.exist?(cache_key)
```

Comment:
We specifically use `CDO.shared_cache` rather than `Rails.cache` here, because the former is shared between threads and the latter isn't.

Reply:
whoa, I missed this in a big way that rails cache is not shared across puma processes and frontends! I wonder if there are other things that we want to move into the shared cache, such as everything in [dashboard/config/initializers/script_preload.rb](https://github.com/code-dot-org/code-dot-org/blob/b6e299a7b092fa2de31970025bd846dc173fffd7/dashboard/config/initializers/script_preload.rb#L11)

I am a surprised all of that can fit into 256MB.

Reply:
Likewise! My original pass at this was done using the Rails cache, until Suresh pointed out that we actually [use an in-memory store for the Rails cache by default](https://github.com/code-dot-org/code-dot-org/blob/5f410aaf17938846af1dd13e125a6097edf4e276/dashboard/config/application.rb#L184) and [don't override that in production](https://github.com/code-dot-org/code-dot-org/blob/5f410aaf17938846af1dd13e125a6097edf4e276/dashboard/config/environments/production.rb#L50-L51).

I think that actually means that we *don't* populate the cache with the Unit cache you linked; my expectation is that `Rails.cache.is_a?(ActiveSupport::Cache::MemoryStore)` is `true` in production

Reply:
I'm trying to be pragmatic here and not sure I have bandwidth right now to pull on this thread of whether there might be a problem with how script preloading is working in production. if the rails cache (only 256MB) is too small for the data we are trying to store in it (with Unit.level_cache likely being the largest), then I could see one of two potential problems:
1. thrashing, likely manifesting in high DB usage
2. actual memory consumed being much higher than 256MB, due to some serious ambiguity around what cache size limit really means when the objects being stored in it are complex activerecord objects 

I would say that problem 1, if it occurs, will likely quickly point us to this codepath. problem 2 may be harder to track down, so perhaps we can just keep in mind that if FE memory usage continues to be high and hard to diagnose, then maybe moving all the script/unit caches from rails cache into shared cache might be a good speculative fix.

I did go a short way down the path of trying to see how much data Unit.level_cache might be putting into the rails cache, and unfortunately the Unit.level_cache command appeared to hang indefinitely for me when run on rails console / production-console, which seems concerning.

Reply:
one additional thought here: according to my notes on https://github.com/code-dot-org/code-dot-org/pull/53398, skipping script preload in dashboard-console in production saves 1.6GB of memory usage. multiply that by the number of puma workers, and there's a good chance you could save that much memory per production FE by moving everything loaded by script_preload.rb from rails cache to shared cache.

Reply:
it won't be quite that simple though, because we store a copy of the cached data rather than accessing the cache on an as-needed basis: https://github.com/code-dot-org/code-dot-org/blob/0148a5e1a6352c8ddbce4bf0be013ff0ce00ee1a/dashboard/app/models/unit.rb#L441-L444

Reply:
Just for my own edification: are there specific issues you're trying to address with this investigation? Is the size of the Rails cache or the current state of frontend memory usage causing us problems somewhere?

Reply:
The goal of this PR is not to reduce frontend memory usage; the goal of this PR is to reduce the number of requests we have to make to S3 when a new frontend starts up

Reply:
rereading this PR, I see that the issue you're trying to solve has to do with request queues filling up, so perhaps there is no reason to worry about production FE memory usage.

Reply:
More specifically, I'd say there are two relevant changes included in this PR: switching from querying the whole bucket at once to querying for individual objects, and switching from an in-memory cache to a shared cache. The first is beneficial because HEAD queries are a lot more efficient in S3 than LIST queries (particularly as this bucket continues to grow), and the second beneficial because it means we only need to make those requests once per server rather than once per thread. In both cases, our goal is to optimize response time; I do expect we will see some reduced memory usage as a side effect, but that's just a bonus.

Reply:
Makes sense! Let's not worry about changing anything about script cache unless production FE memory usage becomes an issue.